### PR TITLE
mds: add performance counter for cap messages

### PR DESCRIPTION
**Pull request by kotreshhr**
Signed-off-by: Kotresh HR <khiremat@redhat.com>
Fixes: https://tracker.ceph.com/issues/47102

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by kotreshhr** [comment]
@batrick @ukernel was the intention to add counters at following specific ops ?

```
enum {
        CEPH_CAP_OP_GRANT,         /* mds->client grant */
        CEPH_CAP_OP_REVOKE,        /* mds->client revoke */
        CEPH_CAP_OP_TRUNC,         /* mds->client trunc notify */
        CEPH_CAP_OP_EXPORT,        /* mds has exported the cap */
        CEPH_CAP_OP_IMPORT,        /* mds has imported the cap */
        CEPH_CAP_OP_UPDATE,        /* client->mds update */
        CEPH_CAP_OP_DROP,          /* client->mds drop cap bits */
        CEPH_CAP_OP_FLUSH,         /* client->mds cap writeback */
        CEPH_CAP_OP_FLUSH_ACK,     /* mds->client flushed */
        CEPH_CAP_OP_FLUSHSNAP,     /* client->mds flush snapped metadata */
        CEPH_CAP_OP_FLUSHSNAP_ACK, /* mds->client flushed snapped metadata */
        CEPH_CAP_OP_RELEASE,       /* client->mds release (clean) cap */
        CEPH_CAP_OP_RENEW,         /* client->mds renewal request */
};
```

**Comment by batrick** [comment]
> @batrick @ukernel was the intention to add counters at following specific ops ?
> 
> ```
> enum {
>         CEPH_CAP_OP_GRANT,         /* mds->client grant */
>         CEPH_CAP_OP_REVOKE,        /* mds->client revoke */
>         CEPH_CAP_OP_TRUNC,         /* mds->client trunc notify */
>         CEPH_CAP_OP_EXPORT,        /* mds has exported the cap */
>         CEPH_CAP_OP_IMPORT,        /* mds has imported the cap */
>         CEPH_CAP_OP_UPDATE,        /* client->mds update */
>         CEPH_CAP_OP_DROP,          /* client->mds drop cap bits */
>         CEPH_CAP_OP_FLUSH,         /* client->mds cap writeback */
>         CEPH_CAP_OP_FLUSH_ACK,     /* mds->client flushed */
>         CEPH_CAP_OP_FLUSHSNAP,     /* client->mds flush snapped metadata */
>         CEPH_CAP_OP_FLUSHSNAP_ACK, /* mds->client flushed snapped metadata */
>         CEPH_CAP_OP_RELEASE,       /* client->mds release (clean) cap */
>         CEPH_CAP_OP_RENEW,         /* client->mds renewal request */
> };
> ```

I don't think soi but also including those ops sounds very useful!

File: src/mds/MDSRank.cc [hunk 1]
```diff
@@ -3254,6 +3254,12 @@ void MDSRank::create_logger()
                             "exi", PerfCountersBuilder::PRIO_INTERESTING);
     mds_plb.add_u64_counter(l_mds_imported_inodes, "imported_inodes", "Imported inodes",
                             "imi", PerfCountersBuilder::PRIO_INTERESTING);
+    mds_plb.add_u64_counter(l_mdss_handle_client_caps, "handle_client_caps",
+                           "Client caps msg", "hcc", PerfCountersBuilder::PRIO_INTERESTING);
```
Comment (ukernel) [inline src/mds/MDSRank.cc:3258]: add counter for both caps messages from client and caps message mds sends. also add counter for client caps messages with dirty caps and mds cap message that revoke caps.

**Comment by ukernel** [comment]
>         CEPH_CAP_OP_RENEW,         /* client->mds renewal request */



> @batrick @ukernel was the intention to add counters at following specific ops ?
> 
> ```
> enum {
>         CEPH_CAP_OP_GRANT,         /* mds->client grant */
>         CEPH_CAP_OP_REVOKE,        /* mds->client revoke */
>         CEPH_CAP_OP_TRUNC,         /* mds->client trunc notify */
>         CEPH_CAP_OP_EXPORT,        /* mds has exported the cap */
>         CEPH_CAP_OP_IMPORT,        /* mds has imported the cap */
above are mds only


>         CEPH_CAP_OP_UPDATE,        /* client->mds update */
client only

>         CEPH_CAP_OP_DROP,          /* client->mds drop cap bits */
 not used

>         CEPH_CAP_OP_FLUSH,         /* client->mds cap writeback */
client only
>         CEPH_CAP_OP_FLUSH_ACK,     /* mds->client flushed */\
mds only 
>         CEPH_CAP_OP_FLUSHSNAP,     /* client->mds flush snapped metadata */
client only
>         CEPH_CAP_OP_FLUSHSNAP_ACK, /* mds->client flushed snapped metadata */
mds only


>         CEPH_CAP_OP_RELEASE,       /* client->mds release (clean) cap */
not used
>         CEPH_CAP_OP_RENEW,         /* client->mds renewal request */
not used
> };
> ```

File: src/mds/MDSRank.cc (hunk 1 above)
Comment (kotreshhr) [reply src/mds/MDSRank.cc:3260]: Done

File: src/mds/Migrator.cc [hunk 2]
```diff
@@ -3591,6 +3595,8 @@ void Migrator::handle_export_caps(const cref_t<MExportCaps> &ex)
     return;
   }
 
+  if (mds->logger) mds->logger->inc(l_mdss_handle_export_caps);
+
```
Comment (ukernel) [inline src/mds/Migrator.cc:3599]: remove all change in this file. they are not cap message

File: src/mds/Locker.cc [hunk 3]
```diff
@@ -2393,6 +2393,7 @@ int Locker::issue_caps(CInode *in, Capability *only_cap)
 
       int op = (before & ~after) ? CEPH_CAP_OP_REVOKE : CEPH_CAP_OP_GRANT;
       if (op == CEPH_CAP_OP_REVOKE) {
+        if (mds->logger) mds->logger->inc(l_mdss_ceph_cap_op_revoke);
```
Comment (ukernel) [inline src/mds/Locker.cc:2396]: also add a counter for op_grant

Comment (kotreshhr) [reply src/mds/Locker.cc:2398]: Done

File: src/mds/Migrator.cc (hunk 2 above)
Comment (kotreshhr) [reply src/mds/Migrator.cc]: Done

**Comment by kotreshhr** [comment]
@ukernel I have added for all the mds cap ops except CEPH_CAP_OP_EXPORT,  CEPH_CAP_OP_IMPORT which has it's usage in MDCache.cc in do_cap_import and handle_cache_rejoin_ack respectively. Do we need counters for those ?

File: src/mds/Locker.cc [hunk 4]
```diff
@@ -2393,10 +2395,13 @@ int Locker::issue_caps(CInode *in, Capability *only_cap)
...
```
Comment (ukernel) [inline src/mds/Locker.cc:2398]: else 
        if (mds->logger) mds->logger->inc(l_mdss_ceph_cap_op_grant);

**Review by ukernel — APPROVED** [review]

**Pull request merged**